### PR TITLE
Support ruby 3

### PR DESCRIPTION
**Pull request by Goltergaul**
I was playing around with how to make receptacle ruby 3 compatible and came up with this. This as WIP PR for early feedback. See the test on how the new wrapper interface is working. Let me know what you think

new functionality is implemented in this file: [lib/receptacle.rb](https://github.com/andreaseger/receptacle/pull/14/files#diff-4025326cfd3158ed4d4f02a49a9f04f12bdd8af98341a13fcec5ea595d128077) see this spec for how the wrapper work now: [spec/receptacle_wrapper_spec.rb](https://github.com/andreaseger/receptacle/pull/14/files#diff-7fee6214397814f9976b5366a0d55ed0eaaeeebc4329e989e4f6620f8a2ebf99) Pleas ignore the rest of the changes. didn't clean up the rest of the code, was in a hurry finishing this up at doni

### CODE REVIEW

**Comment by Goltergaul** [comment]
new functionality is implemented in this file: lib/receptacle.rb see this spec for how the wrapper work now: https://github.com/andreaseger/receptacle/pull/14/files#diff-7fee6214397814f9976b5366a0d55ed0eaaeeebc4329e989e4f6620f8a2ebf99

File: test/fixture.rb [hunk 1]
```diff
@@ -89,14 +97,14 @@ def after_b(return_value, array)
 
     class BeforeAfterWithStateC
       # :c
-      def before_c(string:)
+      def before_c(a, string:)
```
Comment (TheWudu) [inline test/fixture.rb:100]: what is a ?

File: spec/test_repositories.rb [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        number
+      end
+    end
+  end
+
+  module Wrapper
+    class Add5ToNumber
+      def find(number, unscoped:)
+        yield(number + 5, unscoped: unscoped)
+      end
```
Comment (TheWudu) [inline spec/test_repositories.rb:17]: this somehow is complementary to the below wrappers with before_<method_name> + after_<method_name>

File: lib/receptacle.rb [hunk 3]
```diff
@@ -1,14 +1,47 @@
...
+        else
+          @strategy
+        end
+      end
+
+      def with_wrappers(wrappers, method_name, *args, **kwargs, &block)
+        next_wrapper = wrappers.shift
+        if next_wrapper
+          next_wrapper.new.public_send(method_name, *args, **kwargs) do |*sub_args, **sub_kwargs|
+            with_wrappers(wrappers, method_name, *sub_args, **sub_kwargs, &block)
```
Comment (TheWudu) [inline lib/receptacle.rb:35]: how do you do the before/after now ?

File: spec/test_repositories.rb (hunk 2 above)
Comment (Goltergaul) [reply spec/test_repositories.rb:17]: you would do it like this, e.g. for sending an update message:

```ruby
let(:add5_wrapper) do
  Class.new do
    def update(id:, changeset:)
      before = Repositories::Foo.find(id: id)
      after = yield(id: id, changeset: changeset)
      Messages.publish(build_update_message(before, after)
    end
  end
end
```

File: test/fixture.rb (hunk 1 above)
Comment (Goltergaul) [reply test/fixture.rb:100]: ignore the minitests, those are old tests that i'll delete

File: lib/receptacle.rb (hunk 3 above)
Comment (Goltergaul) [reply lib/receptacle.rb:35]: you would do it like this, e.g. for sending an update message:

```ruby
let(:add5_wrapper) do
  Class.new do
    def update(id:, changeset:)
      before = Repositories::Foo.find(id: id)
      after = yield(id: id, changeset: changeset)
      Messages.publish(build_update_message(before, after)
    end
  end
end
```

File: spec/test_repositories.rb (hunk 2 above)
Comment (Goltergaul) [reply spec/test_repositories.rb:17]: its before and after in one. yield gives you the "after"

File: lib/receptacle.rb (hunk 3 above)
Comment (Goltergaul) [reply lib/receptacle.rb:35]: its basically like a rack middleware iirc

**Review by ghsmaniotto — COMMENTED** [review]
I like this new way of using wrappers and strategies, it's a good solution for allowing ruby 3, in my POV!

**Review by thoesl — COMMENTED** [review]
Disclaimer: I think I've only once worked with wrappers so far

But still, I like the new way that is proposed here. imho it's better understandable what's going on and when what happens. Of course it will take a little bit to get used to it but overall it feels less magic than before.

File: lib/receptacle.rb (hunk 3 above)
Comment (Goltergaul) [reply lib/receptacle.rb:35]: in comparison with the old implementation:

```ruby
class MessagesExistingInterface
  def before_update(**args)
    store_before_operation(args)
    { **args }
  end

  def after_update(model, _args)
    changed_attributes = Helper::ChangedAttributes.build(model,
                                                         @before_operation_model)
    return model if changed_attributes.empty?

    MessagePublisher.publish(event:              :updated,
                             entity:             model,
                             changed_attributes: changed_attributes,
                             opts:               opts)
    model
  end

  private

  def store_before_operation(args, model = nil)
    @before_operation_model = model || find_entity(**args) # find_entity has to be defined in service wrapper
  rescue NoMethodError
    nil
  end
end

class MessagesNewInterface
  def update(id:, changeset:)
    before_model = find_entity(id: id)
    yield(id: id, changest: changeset).tap do |model|
      changed_attributes = Helper::ChangedAttributes.build(model, before_model)
      return model if changed_attributes.empty?

      MessagePublisher.publish(event: :updated,
                                                            entity:             model,
                                                            changed_attributes: changed_attributes)
      model
    end
  end
end
```

**Comment by Goltergaul** [comment]
closing this now. will reopen with a final one

**Pull request closed without merging**